crates/dht/src/dht.rs

```diff
@@ -0,0 +1,425 @@
+            self.partition.disc_top_hash(arc_set, store).await?;
+
+        Ok(DhtSnapshot::Minimal {
+            disc_top_hash,
+            disc_boundary,
+            ring_top_hashes: self.partition.ring_top_hashes(arc_set),
+        })
+    }
+
+    /// Handle a snapshot from another DHT model.
+    ///
+    /// This is a two-step process. First the type of the incoming snapshot is checked and a
+    /// snapshot of the same type is computed. Secondly, the two snapshots are compared to determine
+    /// what action should be taken next.
+    ///
+    /// The state flow is as follows:
+    /// - If the two snapshots are identical, the function will return [DhtSnapshotNextAction::Identical].
+    /// - If the two snapshots cannot be compared, the function will return [DhtSnapshotNextAction::CannotCompare].
+    ///   This can happen if the time slices of the two DHTs are not aligned or one side is following
+    ///   a different flow to what we're expecting.
+    /// - If the snapshots are different, the function will return [DhtSnapshotNextAction::NewSnapshot]
+    ///   with a more detailed snapshot of the DHT model.
+    /// - When the most detailed snapshot type is reached, the function will return [DhtSnapshotNextAction::NewSnapshotAndHashList]
+    /// - The new snapshot from [DhtSnapshotNextAction::NewSnapshotAndHashList] should be sent to the
+    ///   other party so that they can compare it with their own snapshot and determine which op hashes
+    ///   they need to fetch.
+    /// - On the final comparison step, the function will return [DhtSnapshotNextAction::HashList]
+    ///   with a list of op hashes. This list should be sent to the other party so that they can fetch
+    ///   any missing ops.
+    ///
+    /// Notice that the final step would require re-computing the most detailed snapshot type. This
+    /// is expensive. To avoid having to recompute a snapshot we've already computed, the caller
+    /// MUST capture the snapshot from [DhtSnapshotNextAction::NewSnapshot] when it contains either
+    /// [DhtSnapshot::DiscSectorDetails] or [DhtSnapshot::RingSectorDetails]. This snapshot can be
+    /// provided back to this function in the `our_previous_snapshot` parameter. In all other cases,
+    /// the caller should provide `None` for `our_previous_snapshot`.
+    ///
+    /// Note also that there are two possible routes through the comparison process. The first is
+    /// when the historical disc mismatches, the second is when the recent rings mismatch. The
+    /// historical disc mismatch is prioritised, so if a mismatch is detected there then the sync
+    /// process will resolve that. Otherwise, the recent rings mismatch will be resolved. That means
+    /// that it may take up to two rounds of sync to resolve all mismatches. Of course, both the
+    /// disc and the rings must be considered a moving target so it cannot be assumed that 2 rounds
+    /// are actually enough to resolve all mismatches.
```

Comment:
Is the idea that, when we complete a disc mismatch round, we go into a second round with no delay to see if we need to resolve any ring mismatches?

Reply:
Not unless we make a modification to not check the `disc` here by keeping our minimal snapshot around and using it to create a ring diff or something like that because until the ops have been fetched, we'll like get another ring diff that's the same as we just got if we go straight back into another sync. I was just thinking that would have to happen on the next gossip round.

Doing it how I was thinking means that we could get stuck syncing historical data if more keeps showing up and not get to recent data... Until that data becomes historical. I'm relying on some "reasonable" behavior here where historical data showing up isn't a constantly occurring thing.